File: rl/src/panenka/actions.py
# Discrete action space, mirroring web/src/lib/engine/actions.ts:
# index = (move_x + 1) * 6 + (move_y + 1) * 2 + kick, with move_x, move_y in {-1, 0, 1}.
import jax.numpy as jnp

ACTION_COUNT = 18
IDLE = 8  # no movement, no kick

def encode_action(move_x, move_y, kick):
    return (move_x + 1) * 6 + (move_y + 1) * 2 + kick


def decode_action(index):
    """Returns (move, kick): move has shape (..., 2), kick is a bool array."""
    index = jnp.asarray(index)
    move = jnp.stack([index // 6 - 1, index // 2 % 3 - 1], axis=-1)
    return move, index % 2 == 1

File: web/src/lib/engine/actions.ts
import type { Action } from './types';

// Discrete action space: one compact index per action, for network input and RL.
// index = (moveX + 1) * 6 + (moveY + 1) * 2 + kick
export const ACTION_COUNT = 18;

export const encodeAction = (a: Action): number => (a.moveX + 1) * 6 + (a.moveY + 1) * 2 + (a.kick ? 1 : 0);

export const decodeAction = (i: number): Action => ({
  moveX: (Math.floor(i / 6) - 1) as Action['moveX'],
  moveY: ((Math.floor(i / 2) % 3) - 1) as Action['moveY'],
  kick: i % 2 === 1,
});

export const IDLE: Action = { moveX: 0, moveY: 0, kick: false };
